### PR TITLE
Purchases: Redirect in `CancelPrivateRegistration` and `EditCardDetails` for invalid data

### LINKED ISSUE #2519: Purchases: missing page redirects
@gziolo implemented some page redirects while working on #275, but the following are still missing:

- [x] Perform redirect when user can't access Confirm Cancel Purchase page
- [ ] Perform redirect when user can't access Cancel Private Registration page
- [ ] Perform redirect when user can't access Edit Card Details page

Related to #1407 and #1870.

### PR DESCRIPTION
**Pull request by drewblaisdell**
This PR adds logic to `CancelPrivateRegistration` and `EditCardDetails` to redirect to `/purchases` if an invalid purchase or card is present.

- In `CancelPrivateRegistration`, this means that the purchase must exist and have a private registration.
- In `EditCardDetails`, this means that the purchase and card must exist.

Fixes #2519 and #1407.

**Testing**
`CancelPrivateRegistration`
- Visit `/purchases` and click on a domain that has a private registration.
- Click 'Cancel Private Registration'.
- Refresh the page and assert that you are not redirected to `/purchases` once the data has loaded.
- Update the site slug or purchase ID in the URL to a random string/number and visit this URL.
- Assert that you are redirected to `/purchases` once the data has loaded.

`EditCardDetails`
- Visit `/purchases` and click on a purchase that was made with a credit card.
- Click 'Edit Payment Method'.
- Assert that you are taken to `EditCardDetails` and that a credit card form is displayed once the data has loaded.
- Update the site slug, purchase ID (the first number) or card ID (the second number) in the URL to a random number and visit this URL.
- Assert that you are redirected to `/purchases` once the data has loaded.

`CancelPurchase`
- Visit `/purchases` and click on a purchase that can be canceled, e.g. any product within its refund window or any subscription with auto-renew enabled.
- Click the cancel link, e.g. 'Cancel Subscription and Refund' or similar.
- Refresh this page and assert that it loads properly.
- Update the site slug or purchase ID in the URL to a random string/number and visit this URL.
- Assert that you are redirected back to `/purchases` once the data has loaded.

`ConfirmCancelDomain`
- Visit `/purchases` and click on a domain within its refund window.
- Click 'Cancel Domain and Refund'.
- Click the 'Cancel Domain and Refund' button.
- Refresh the page and assert that it loads properly.
- Update the site slug or purchase ID in the URL to a random string/number and visit this URL.
- Assert that you are redirected back to `/purchases` once the data has loaded.

- [x] Code review
- [x] Product review

### CODE REVIEW

**Comment by drewblaisdell** [comment]
I'm going to update this to fix #1407 as well.

**Comment by spncrb** [comment]
LGTM :+1: 

Noticed this while testing: https://github.com/Automattic/wp-calypso/issues/2519 , but it is already filed so we good...

File: client/me/purchases/payment/edit-card-details/index.jsx [hunk 1]
```diff
@@ -86,6 +92,10 @@ const EditCardDetails = React.createClass( {
 		this.setState( { form: this.formStateController.getInitialState() } );
 	},
 
+	componentWillReceiveProps( nextProps ) {
+		this.redirectIfDataIsInvalid( nextProps );
```
Comment (scruffian) [inline client/me/purchases/payment/edit-card-details/index.jsx]: Not related to this PR at all, but do you think we should add a call to `recordPageView` here, like we do in `client/me/purchases/cancel-private-registration/index.jsx`?

**Comment by scruffian** [comment]
Code :+1:

File: client/me/purchases/payment/edit-card-details/index.jsx (hunk 1 above)
Comment (drewblaisdell) [inline client/me/purchases/payment/edit-card-details/index.jsx]: Good catch. I opened an issue for this here: https://github.com/Automattic/wp-calypso/issues/4590

**Comment by drewblaisdell** [comment]
Thanks @spncrb / @scruffian.

**Pull request merged**